### ISSUE TITLE
Eliminate `factory` modules

### ISSUE THREAD

**Issue by mhsmith** [beeware/toga#2687]
The sub-module structure of each backend is identical, so that can already be considered a well-defined API between the interface and backend layers. Putting the "factory" namespace in between them adds complexity for no benefit, and also forces all the sub-modules to be imported whether the app uses them or not (#2547).

For example, where we currently write:
```
self.factory.Camera
```

We could instead have:
```
import_backend("hardware.camera.Camera")
```

Which on macOS would translate into an import of `toga_cocoa.hardware.camera`, and a getattr of `Camera`.

For previous discussion, start at https://github.com/beeware/toga/pull/2584#pullrequestreview-2092686462.

**Comment by freakboy3742** [comment]
As noted in [this comment](https://github.com/beeware/toga/pull/2584#issuecomment-2148981111), it may be preferable to expose this API as `import_backend("hardware").camera.Camera` or `import_backend("hardware.camera").Camera`; that makes import_backend a relatively lightweight wrapper around `import_module`, but injecting the platform name, and yielding a module.

A related thought: this mechanism could also be used to provide a plugin interface for third-party widgets to provide platform-specific implementations. Consider the case of a third party widget that has native platform implementations - the widget could register as having providing a macOS implementation, then the interface use a standard `import_backend` API to access the platform backend. That avoids third-party widget needing to replicate the "if sys.platform ..." logic that would be the core of `import_backend`, and should be able to do this as a fallback if an `AttributeError` is raised on a first-pass lookup on the built-in backend.

**Comment by HalfWhitt** [comment]
I've been looking into this (and #2928), and I've noticed an inconsistency.

Remember that `_IMPL_NAME` property I stuck on widget classes in my first pass on widget initialization in #2942, as a way of centralizing the implementation-finding logic, yet still keeping the same implementation if a user subclasses an existing Toga widget? `Window` uses [essentially the same thing](https://github.com/beeware/toga/blob/1c8b7d41dbf2ad171a882388005960ddd434a5c4/core/src/toga/window.py#L133) with its `_WINDOW_CLASS` attribute. And `StatusIcon` uses the [same sort of dynamic lookup](https://github.com/beeware/toga/blob/1c8b7d41dbf2ad171a882388005960ddd434a5c4/core/src/toga/statusicons.py#L24) to get its implementation, but simply uses the current object's class name... which would break a user-made subclass.

I'm not entirely sure yet what, if anything, I want to do about either of those, but it seems they could be part of whatever this implementation-fetching redesign ultimately looks like.

Seeing as how this feature will change how users create and subclass widgets — which #2928 already does too — perhaps it would be best to get this in before 0.5.0, so there's only one churn instead of two.

**Comment by freakboy3742** [comment]
> I'm not entirely sure yet what, if anything, I want to do about either of those, but it seems they could be part of whatever this implementation-fetching redesign ultimately looks like.

Agreed - my initial impression would be that using the `_create()` approach from #2942 would at least be consistent.

> I share your suspicion that this and #2928 are closely linked. Since adding them will change how users can create and subclass custom widgets — which #2942 already does too — perhaps it would be best to get these in before 0.5.0, so there's only one churn instead of two.

Agreed. The combination with this ticket also highlights that although #2928 is framed in terms of "widgets", there's a need to register other types of objects - windows, icons, hardware, and potentially more.

**Comment by HalfWhitt** [comment]
> Agreed - my initial impression would be that using the `_create()` approach from #2942 would at least be consistent.

Sounds reasonable. Even if things ultimately become something different, at least doing that as a first pass means it'll be easier to change everything the same way.

Another thing, too, as I ponder and sketch out how all this fits together — in what form do we expect people to be creating and providing Toga additions? Toga itself is a separate distribution package (and import package) for core and for each backend. That's great for flexibility and only installing what you need, but it's also pretty complex as a minimum requirement for someone to start trying out their custom widget, especially if they only have one or two backends implemented. If the registration mechanism allows one to specify the root of each backend, then someone could provide separate packages *or* simple submodules inside their package.

Also... It would go against the precedent set with image plugins and backends themselves, but I keep wondering if it might be worth using a registration mechanism along the lines of what Python does with ABCs — you create your own sequence type and then do `Sequence.register(my_type)`. It's explicit in the code itself, and it can be done / tested out even in code that isn't part of an installed package and thus lacks entry points.

> Agreed. The combination with this ticket also highlights that although #2928 is framed in terms of "widgets", there's a need to register other types of objects - windows, icons, hardware, and potentially more.

I suspect most subclassing of Window will be to customize or add behavior at the core level, but I suppose you never know... I think widgets and hardware should probably be the priority, as far as registering custom implementations goes.

**Comment by freakboy3742** [comment]
> Another thing, too, as I ponder and sketch out how all this fits together — in what form do we expect people to be creating and providing Toga additions? 

Good question - I don't know for sure. A single installable package is easy to set up and explain, but complicates the process of defining and installing dependencies. Multiple installable packages (mirroring the Toga repo setup) is the most robust, but a lot harder to set up. We may need to provide template repositories for bootstrapping new widgets.

> Also... It would go against the precedent set with image plugins and backends themselves, but I keep wondering if it might be worth using a registration mechanism along the lines of what Python does with ABCs — you create your own sequence type and then do `Sequence.register(my_type)`. It's explicit in the code itself, and it can be done / tested out even in code that isn't part of an installed package and thus lacks entry points.

AFAIK, that's mostly an accident of history. The metadata associated with extras and plugins has really only been well formalised since 3.8 or thereabouts; so a lot of patterns in the CPython codebase are born out of necessity because the metadata couldn't be relied on.

**Comment by HalfWhitt** [comment]
> A single installable package is easy to set up and explain, but complicates the process of defining and installing dependencies. Multiple installable packages (mirroring the Toga repo setup) is the most robust, but a lot harder to set up. We may need to provide template repositories for bootstrapping new widgets.

Perhaps even a script that generates a skeleton, à la what what Briefcase can do for a new project.

> AFAIK, that's mostly an accident of history. The metadata associated with extras and plugins has really only been well formalised since 3.8 or thereabouts; so a lot of patterns in the CPython codebase are born out of necessity because the metadata couldn't be relied on.

Huh, I hadn't put two and two together there. I still don't really have entry points well stored in my head — despite having added them for the image plugins — but it's probably just because they're still new to me.
 
---

Oh, hey. @t-arn , just giving you a heads up about all this. With this issue, #2928 (third-party widget registration), and #2942 (re-ordered widget initialization), the changes for third-party widget definition in 0.5.0 are still in flux, but are probably going to affect your taTogaLib.

**Comment by t-arn** [comment]
@HalfWhitt Thanks for the warning. It's always good to know when breaking changes are around the bend.

**Comment by corranwebster** [comment]
Since I've been looking at this code, it might be worth describing how this was handled in the Enthought Tool Suite: we had a lot of similar issues, although the backends weren't platform-dependent, we did need to decide on a Qt backend vs a Wx backend.  The code went through a *lot* of changes over the years (the initial implementations were in the mid-2000's) but what we ended up with was this: https://github.com/enthought/pyface/blob/main/pyface/base_toolkit.py (A 'toolkit' is roughly analogous to a 'backend').

The architecture was a little different (common functionality was provided by a mixin class along with an interface (effectively a `typing.Protocol` before they existed) that needed to be implemented, and concrete implementations were used directly). So when you did the equivalent of `from toga import FooWidget` (eg. `from pyface.api import FooWidget`) you got a lazily-loaded *implementation* that declared that it adhered to a *protocol*. So there's not a 1-to-1 mapping with the Toga architecture, but I think the ideas may still be relevant.

The other thing we did which might be worth considering, was that part of the selection process was testing to see if the backend would work. This was a bit rudimentary: basically we tried to import the underlying libraries as part of the entrypoint import, and if the entrypoint load failed we tried to sort out why. If I were implementing this from scratch today, I might add a function or method to the entrypoint object that could give a better answer about whether things are supported. So your loading code might look like:
``` python
toga_backends = find_backends()
backends = [entrypoint.load() for entrypoint in toga_backends]
supported_backends = [backend.is_supported() for backend in backends]
if len(supported_backends) == 1:
    return supported_backend
else:
    #fail appropriately
```
The backend's `is_supported` could then do whatever it wanted - maybe just check the platform, or maybe do some deeper checking to make sure everything works (eg. is Gtk/PySide6/Textual importable?)

When it came to third-party registration of widgets, we did have a mechanism to allow multiple packages to supply widgets for a given implementation, but we never really used that. Rather the base toolkit mechanism was generic, so when we had packages that provided extra widgets (eg. Enable/Chaco and MayaVi which provided 2D and 3D visualization), they simply used the same `Toolkit` class, but pointed at a different entrypoint,(ie. rather than `pyface.toolkits`, look in `enable.toolkits`) but they used the same *names* for the entrypoints together with a global setting that could be looked up (`ETSConfig.toolkit`) so if a toolkit was already selected by one package, the same one would be used by all.

**Comment by freakboy3742** [comment]
Thanks for those details - it's always helpful to see how others have tackled similar problems.

Regarding entry point specifically - I guess the question is whether it's the responsibility of the backend to be able to report that it is supported, or whether that could be part of the entry point itself - if you don't support GTK, don't register a GTK entry point. That means backends would be registering more entry points; but they're not expensive to set up or interrogate, so I don't think there's a huge problem going down the explicit path.

**Comment by corranwebster** [comment]
> Regarding entry point specifically - I guess the question is whether it's the responsibility of the backend to be able to report that it is supported, or whether that could be part of the entry point itself - if you don't support GTK, don't register a GTK entry point.

Yes, part of the problem that we had was that everything was in one package (for historical reasons) and that meant we couldn't simply declare a dependency of the package on a particular library like GTK.

>That means backends would be registering more entry points; but they're not expensive to set up or interrogate, so I don't think there's a huge problem going down the explicit path.

So you're thinking a "factory" or it's replacement might look something like:
``` python
class Factory:
    def __init__(self, backend_name):
        self.backend_name = backend_name
        self._entrypoints = {
            entrypoint.name: entrypoint
            for entrypoint in entrypoints.group(f"toga.backend.{backend_name}")
        }

    def __getattr__(self, name):
        if name in self._entrypoints:
            value = self._entrypoints[name].load()
            setattr(self, name, value)
            return value
        else:
            raise AttributeError(name)
```

**Comment by freakboy3742** [comment]
> So you're thinking a "factory" or it's replacement might look something like:

Possibly. I haven't given this a huge amount of thought beyond what I've posted historically in this channel; essentially, the use cases I can see that we need to capture are:

* Core widgets and implementations
* A third party implementation of a widget defined by core, but not implemented in core (e.g., adding Table for iOS right now)
* A third party widget that provides a new widget with platform-specific backends (e.g., adding a "BitmapView" widget)
* A "fourth" party implementation of a widget provided by a third-party (e.g., adding an iOS backend for the BitmapView from the previous example)
* A third party "composite" widget (i.e., a widget made up of other widgets) - e.g., toga-chart

I chose BitmapView specifically because [I *have* some (very out of date) example code for this](https://github.com/beeware/yorkshire4/tree/main/yorkshire4/widgets) that needs to be pulled into a standalone repo; but it could be anything - even a "reimplementation of toga.Button" would be a candidate for test purposes.

The tl;dr is that this is a project that requires some design work.

**Comment by corranwebster** [comment]
So I sat down yesterday and had a good hard think about how all this might work, and came up with the following design, which leans on the entry point system heavily:

## Backend Selection

Toga discovers available backends via Python's entry point mechanism. Each backend advertises that it provides a backend by setting an entry-point in the `toga.core.backends` group in its package metadata.

Normally there should be exactly one backend installed in an application's virtual environment, and so Toga will use that. However it is possible that there might be multiple valid backends installed in an environment, such as when a developer is working on Toga itself. For example on Linux there could be any out of the Gtk, Qt, Textual and Web backends installed and working in a given environment.

When faced with multiple backends, Toga will look for the `TOGA_BACKEND` environment variable, and use the backend named there. Otherwise it will raise an Exception.

Once the backend is selected, the name is stored in a known place (eg. `toga.backend`), ideally in a lazy way so that when it is requested the first time it triggers the selection process..

## Implementation Objects and the Factory

Once the backend is known, the implementations of the widgets and other backend features are also provided via entrypoints in the backend's package that contribute to the `toga.core.backend.{backend}` group. So, for example all the iOS backend widgets are contributed to the `toga.core.backend.{backend}` group.

The `toga.platform.factory` object is responsible for lazily loading the backend objects as they are needed. Widget implementation classes are available as `factory.<widget>` (so for example, the implementation of the `LineEdit` widget for the active backend can be accessed as `factory.LineEdit`).

The factory is an instance of a general `Factory` class that is given a project name (the default is "core") and looks up the current backend name from `toga.backend`, and then lazily loads from the entrypoint group `toga.<project_name>.backend.<backend_name>`, so if the current backend is `toga_cocoa` then:

``` python
my_factory = Factory("my_project")
my_factory.MyWidget
```
will look for an entrypoint with name "MyWidget" in the entrypoint group `toga.my_project.backend.toga_cocoa` and then call the entrypoint's `load()` method to get the implementation class object. If no such object is found, it will raise an error. It then adds the object as an attribute so it doesn't need to load the entrypoint multiple times.

# Use Cases

This scheme is designed to support a number of use cases.

## Adding a New Widget to Core Toga

The developer needs to write the interface object and the concrete interface classes in each backend. For each backend, the developer needs to add an entry point of the form `<Widget class name> = <implementation module>:<Widget class name>` to the `toga.core.backend.<backend name>`.

So for example, the iOS canvas widget is registered in the `pyproject.toml` for `toga_iOS`.

``` toml
[project.entry-points."toga.core.backend.toga_iOS"]
Canvas = "toga_iOS.widgets.canvas:Canvas"
```

## Adding a Third-Party Implementation for an Unimplemented Core Widget

Some backends may not yet have implementations for all core widgets. In these cases if your application needs this widget, you can write an implementation and make it available via an entry point in your application or library. Add an entrypoint for the widget to you application's `pyproject.toml`.

So say you had implemented a `Table` widget for iOS in a module `my_ios_app.widgets.table` in your application, then you could add the following to your `pyproject.toml`:

``` toml
[project.entry-points."toga.core.backend.toga_iOS"]
Table = "my_ios_app.widgets.table:Table"
```

## Adding a new Backend

To add a new backend for Toga you need to create a project with the backend implementations of the Toga widgets. You then need to add entrypoints for your backend, and then for each object that Toga expects to see in the factory.

So for example, to implement a backend for watchOS, you would add something like the following to the `pyproject.toml`:

``` toml
[project.entry-points."toga.core.backends"]
watchOS = "toga_watchOS"

[project.entry-points."toga.core.backend.watchOS"]
App = "toga_watchOS.app:App"
DocumentApp = "toga_watchOS.app:DocumentApp"
Button = "toga_watchOS.widgets.button:Button"
...
```

It probably makes sense to follow the same module and class layout as other backends, but there is no requirement to do so.

Note that the name of the backend doesn't have to match the name of the platform: in most cases the dependencies for your backend will make it difficult or impossible to install on a platform where it won't work.


## Adding Your Own Custom Widgets

You may find your application or library has a need for a new Widget which isn't available as part of Toga yet. The first step, as with any widget, is writing the interface class; after that you need to write the implementation of the backends you need.

For a simple application with one backend, this may be enough, with your interface code simply importing the implementation object directly in the `create` method:

``` python
from toga import Widget

class MyWidget(Widget):
    def create(self):
        from my_app.backend.my_widget import MyWidget
        self._impl = MyWidget(self)
```

But if you want your application to be cross-platform, you will have multiple implementations for different backends, and you should do the following:

- create a new `Factory` object for your application or library with it's own project name and make that factory object part of your internal API.
- in your `pyproject.toml` add entrypoints for each implementation of the interface of in groups of the form `toga.<project_name>.backend.<backend_name>`.

Particularly complex applications or libraries which are defining many new widgets may find it useful to follow Toga's example and break out the implementations into separate projects to make dependency management easier.

### Example

As a concrete example, assume that you are writing a cross-platform bitmap editing application. In your application, you create a Factory object for your project, passing it the name you want to use for your entrypoint groups (in this case `"my_app"`).

In module :my_app.platform`
``` python
from toga.platform import Factory

factory = Factory("my_app")
```

You then might write the interface for a new `BitmapView` widget. In `my_app.widgets.bitmap_view`:
``` python
from toga import Widget
from my_app.factory import my_factory

class BitmapView(Widget):

    def create(self):
        self._impl = my_factory.BitmapView(self)
```

Then you might write the implementation for macOS. In `my_app.cocoa.widgets.bitmap_view`:
``` python
from toga_cocoa.widgets.base import Widget

class BitmapView(Widget):
    def create(self):
        self.native = ...
```

And another for GTK. In `my_app.gtk.widgets.bitmap_view`:
``` python
from toga_gtk.widgets.base import Widget

class BitmapView(Widget):
    def create(self):
        self.native = ...
```

And then in your `pyproject.toml` you might have something like:
``` toml
[project]
dependencies = ['toga_core']

[project.optional-dependencies]
cocoa = ["toga_cocoa"]
gtk = ["toga_gtk"]

[project.entry-points."toga.my_app.backend.toga_cocoa"]
BitmapView = "my_app.cocoa.bitmap_view:BitmapView"

[project.entry-points."toga.my_app.backend.toga_gtk"]
BitmapView = "my_app.gtk.bitmap_view:BitmapView"
```

Note that instead of the "toga.core.backend..." groups, you are contributing to the "toga.my_app.backend..." groups, matching the project name of the toolkit object

## Adding an Implementation of a Third-Party Widget for a Backend

It may be that someone has written a third-party widget that is useful to you, but not on the platform you need. Assuming that they have followed the pattern of the previous section and created a factory object of their own, then if you write an implementation of the backend in your project you can contribute it to their project's `toga.<project_name>.backend.<backend_name>` entrypoint group.

### Example

Continuing the provious example, assume that someone has a pjorject as described above that has a `BitmapView` widget, but only for macOS and GTK.  If you needed an implementation for Qt, for example, you could write your own implementation, something like this:

In module `another_app.qt.bitmap_view`:
``` python
from toga_qt.widgets.base import Widget

class BitmapView(Widget):
    def create(self):
        self.native = ...
```

And then add it to *your* `pyproject.toml` contributing to *their* entrypoint groups:
``` toml
[project.optional-dependencies]
qt = ["toga_qt"]

[project.entry-points."toga.my_app.backend.toga_qt"]
BitmapView = "another_app.qt.bitmap_view:BitmapView"
```

Once you've done this you can import the interface from their library and use it in your application with the backend you want:

In module `another_app.app`:
```python
from toga import App
from my_app.widgets.bitmap_view import BitmapView

class AnotherApp(App):
    def startup(self):
        bitmap_view = BitmapView()
        ...
```

## Writing Composite Widgets or Reusing Implementations

It might be the case that an existing core widget implementation provides everything you need for another widget.

For example, you may be writing an HTML documentation-browsing widget for you application's help system. For these purposes, the WebView widget's implementation likely provides everything that you need, but you don't need the full power of it. In cases like this, you can get the implementation from the factory and use it in your interface class:

``` python
from pathlib import Path
from toga import Widget
from toga.platform import toolkit

HELP_PATH = Path(...)

class HelpViewer(Widget):
    def __init__(self, topic=None, ...):
        super().__init__(...)
        self.topic = topic

    def create(self):
        self._impl = toolkit.WebView(self)

    @property
    def topic(self):
        return self._topic

    @topic.setter
    def topic(self, value):
        path = HELP_PATH.join(value + ".html")
        content = open(path).read()
        self._impl.set_content(content)
```

## Overriding the Implementation of a Widget

Sometimes you may need to use a different implementation of a widget for your application.

In this case the simplest approach may be a subclass of the original widget which uses your implementation: this is essentially the same process as adding a custom widget.

XXX: not sure the following is a good idea!

It might be that this doesn't work for your needs and you need to use the original widget interface but with a different implementation, for example replacing the implementation with a mock for testing. You can't use the entrypoint mechanism for this: if the factory discovers multiple entrypoints for the same widget in the same backend, it is undefined which one will be chosen.

What you can do is define your own factory object for your app, and contribute your implementations for each backend to that factory as if you were writing implementations for your own custom widget::

In module `my_app.cocoa.line_edit`:
``` python
from toga_cocoa.widgets.line_edit

class MyLineEdit(LineEdit):
    # override with new behaviour
```

And in the `pyproject.toml` have entries like
``` toml
[project.entry-points."toga.my_app.backend.toga_cocoa"]
LineInput = "my_app.cocoa.line_edit:MyLineEdit"
```

And then in your App, or anywhere else you are creating widgets, do something like:
``` python
from toga import App, LineInput
from toga.platform import factory

from my_app.platform import my_factory

class MyApp(App):
    def startup(self):
        # replace core LineInput implementation with custom one
        factory.LineInput = my_factory.LineInput

        line_input = LineInput()  # patched line input
        ...

        # the core LineInput can be restored
        del factory.LineInput

        another_line_input = LineInput()  # unpatched line input
        ...
```

XXX: The following is an even worse idea

In the extreme, you could create your own factory instance and monkeypatch `toga.platform` with your own. In this case you'd need to make sure that you have entrypoints for everything that you're going to use, including ones pointing to the core implementations.

# Design Notes

In comparison to the current system, this makes the following changes:

- the notion of a "platform", or the linking between platforms and backends, is diminished. In practice, most of the time an application will have a well-controlled environment and so there will be exactly one backend. A regular entrypoint search should be enough, augmented by the ability to override with an environment variable to force selection which would be used by developers who happen to have multiple backends in an environment.
- This has the side-effect of making support for new platforms easier: there is no need to update `toga.platforms._TOGA_PLATFORMS` every time a new platform comes along (eg. currently OpenBSD isn't in this list, and isn't listed as an entrypoint for Gtk or Qt backends, but I would expect the GTK or Qt backends to work on a correctly configured system).
- we might want to have a manually curated list of "preferred backends" for each platform, but even them that feels like a problem for Briefcase more than Toga.
- the backend "factory" modules can be dispensed with completely.
- the number of entrypoint searches isn't unreasonable. It should be two in most cases - one to lookup the backend, and one to find all the widget entrypoints. I'm old enough that my instincts are that entrypoints are slow, but I think `importlib.metadata` is significantly more performant than `pkg_resources`.
- rather than using `import_module` we use `EntryPoint.load` which is more flexible and doesn't require exact matching between organization of modules in the backend (particularly important for an app that needs to add a custom widget or two).
- using entry points for widgets, rather than via an ad-hoc registration system decouples things significantly, and doesn't require working out how to run the registration code for a third-party library.

**Comment by mhsmith** [comment]
Thanks for the detailed proposal. Someone will eventually respond to it, but because of holidays it might not be for a few weeks.

**Comment by freakboy3742** [comment]
Overall, this seems like a really solid design - and, ultimately, shouldn't be *that* hard to implement. Nice work! A couple of questions inline about some of the specifics:

> So I sat down yesterday and had a good hard think about how all this might work, and came up with the following design, which leans on the entry point system heavily:
> 
> ## Backend Selection
> 
> Toga discovers available backends via Python's entry point mechanism. Each backend advertises that it provides a backend by setting an entry-point in the `toga.core.backends` group in its package metadata.

To be clear - this is already done (although it's spelled `toga.backends`)
 
> The factory is an instance of a general `Factory` class that is given a project name (the default is "core") and looks up the current backend name from `toga.backend`, and then lazily loads from the entrypoint group `toga.<project_name>.backend.<backend_name>`, so if the current backend is `toga_cocoa` then:
> 
> my_factory = Factory("my_project")
> my_factory.MyWidget
> will look for an entrypoint with name "MyWidget" in the entrypoint group `toga.my_project.backend.toga_cocoa` and then call the entrypoint's `load()` method to get the implementation class object. If no such object is found, it will raise an error. It then adds the object as an attribute so it doesn't need to load the entrypoint multiple times.

The general idea for this makes sense; however, is there value in the extra layer of needing to create an instance of `Factory()`? Why not have a single toga API of `factory(namespace, widget_name)`, and leave it up to Toga's internals to keep a cache of loaded namespaces? Is there anything stateful that can't be retained at the level of Toga itself? 

Also - in terms of namespacing, we've been trying to encourage the use of a `togax` namespace for "unofficial"  third-party widgets; it might make sense to use that naming in the hooks as well. That also gives us an opportunity to actually document the naming convention.

So - we'd use `toga.backend.toga_iOS` for Toga's own objects, and `togax.<project>.backend.toga_iOS` for third-party ones (which also simplifies the need for the `core` part of the namespace for built-in widgets).
 
> # Use Cases
> 
> This scheme is designed to support a number of use cases.

With the exception of the naming issues and question about `factory()` above, this all looks really solid to me. 

The only question I've got is related to performance - whether lazy loading from an entry point for *every* widget will have a notable performance overhead. It *might* be worth hard-coding the "known built-in" widgets and using the lookup scheme when the namespace isn't "toga" or the widget isn't found. 

However, it also might be the exact opposite.  #2547 is tracking the idea of lazy loading widgets so if your app only uses Box Button and label, you don't carry the overhead of loading *all* widgets from the backend. Some experimentation might be called for here.

The one possible hole I can see is what happens if *two* modules define an entry point for the same widget? i.e., if my environment has 2 independent macOS implementations for BitmapView (or, I guess, Button for that matter), how is this identified and resolved? You allude to this in your discussion about subclassing and composite widgets; but it seems like a core property of the loading system itself, so we need to make sure we have a solid answer for this.

> ## Writing Composite Widgets or Reusing Implementations
> 
> It might be the case that an existing core widget implementation provides everything you need for another widget.
> 
> For example, you may be writing an HTML documentation-browsing widget for you application's help system. For these purposes, the WebView widget's implementation likely provides everything that you need, but you don't need the full power of it. In cases like this, you can get the implementation from the factory and use it in your interface class:
> 
> from toga.platform import toolkit

I'm not sure if this is drift from an earlier draft - should this be `factory`? Where does `toolkit` come from?

> ## Overriding the Implementation of a Widget
> 
> Sometimes you may need to use a different implementation of a widget for your application.
> XXX: not sure the following is a good idea!
> XXX: The following is an even worse idea

I think these fall into the category of "interesting crimes that may be possible but that we probably shouldn't encourage" :-) 
 
> # Design Notes
> 
> In comparison to the current system, this makes the following changes:
> 
> * the notion of a "platform", or the linking between platforms and backends, is diminished. In practice, most of the time an application will have a well-controlled environment and so there will be exactly one backend. A regular entrypoint search should be enough, augmented by the ability to override with an environment variable to force selection which would be used by developers who happen to have multiple backends in an environment.
> * This has the side-effect of making support for new platforms easier: there is no need to update `toga.platforms._TOGA_PLATFORMS` every time a new platform comes along (eg. currently OpenBSD isn't in this list, and isn't listed as an entrypoint for Gtk or Qt backends, but I would expect the GTK or Qt backends to work on a correctly configured system).

Agreed these two properties are desirable; the concept of the "platform" is really only useful as a "tiebreak" when there is more than one backend installed - which is *mostly* a problem for developers *of* Toga, rather than end-users.

> * we might want to have a manually curated list of "preferred backends" for each platform, but even them that feels like a problem for Briefcase more than Toga.

... or the `toga` meta-package, which specifies which backend is installed using `; sys_platform == "..."` qualifiers.

**Comment by corranwebster** [comment]
> The general idea for this makes sense; however, is there value in the extra layer of needing to create an instance of `Factory()`? Why not have a single toga API of `factory(namespace, widget_name)`, and leave it up to Toga's internals to keep a cache of loaded namespaces? Is there anything stateful that can't be retained at the level of Toga itself?

The factory object effectively performs entry-point query caching: it does one query for all the entrypoints for `namespace` entrypoint group rather than having to query each entrypoint individually on each load. Note that it doesn't actually *load* the widgets after the queries, it's basically just holding a look-up of "this widget is here" and lazily loads (effectively an import) when it's first used.

So you want that cache *somehow* - it doesn't have to be a class that's holding it.

The other part of the API above was to keep the API similar to what is there now, but it doesn't have to be via attribute lookup. Either the widget needs to be given the backend name somehow, or it needs to be given an object which wraps that.  My feeling is that the latter may be easier to test and mock, for example, but at its core, as long as you have the cache it's all API details.

> The only question I've got is related to performance - whether lazy loading from an entry point for every widget will have a notable performance overhead. It might be worth hard-coding the "known built-in" widgets and using the lookup scheme when the namespace isn't "toga" or the widget isn't found.

I don't think so, this scheme requires only two entry-point queries (which are the potentially slow things):
- a query to find all available backends
- a query to get the locations of the widgets (which you cache in the factory or whatever replaces it's functionality)
Actually loading an entrypoint is essentially a dictionary lookup followed by an import `from foo.bar import baz` which you'd need to do anyway.

Given that you're going to need to do the entrypoint queries anyway, I think the overhead is minimal and you won't win much by having some values hard-coded.

> However, it also might be the exact opposite. https://github.com/beeware/toga/issues/2547 is tracking the idea of lazy loading widgets so if your app only uses Box Button and label, you don't carry the overhead of loading all widgets from the backend. Some experimentation might be called for here.

This proposal does have this lazy-loading advantage: because the imports are only done when the widgets are requested, you're not needing to import potentially heavy-weight widgets like web views if you aren't using them.  In fact, because entrypoints can be split across packages, backend implementations can be split so you don't even have to include the web view in your app's runtime if you don't need it.

> The one possible hole I can see is what happens if two modules define an entry point for the same widget? i.e., if my environment has 2 independent macOS implementations for BitmapView (or, I guess, Button for that matter), how is this identified and resolved? You allude to this in your discussion about subclassing and composite widgets; but it seems like a core property of the loading system itself, so we need to make sure we have a solid answer for this.

So if you had two implementations of a widget (say two implementations of a missing core widget from a backend), then there'll be two values for that widget, which we *can* detect and warn, and then pick one (we have no way of knowing if one or the other is preferable, although there may be some heuristics we could come up with).

The fix for the user is:
- if you control either or both packages, remove the entrypoint (eg. you implemented a missing widget, now core provides it)
- if you're in a situation where you can't get rid of the entrypoints for whatever reason, you can:
  - subclass the widget interface and directly specify the one you really want in your app
  - create your own backend (ie. a bunch of entrypoints, not any actual code) which is mostly a copy the backend that you are using, but specifies which widget implementation you want by its location (eg. `FooWidget = togax_foo.backend.foo_widget:FooWidget` instead of `FooWidget = togax_bar.backend.foo_widget:FooWidget`)

This sort of collision is inherent in any sort of plugin system. The best you can do is have good heuristics and a work-around for when the heuristics fail.

That plus perhaps encourage people who are writing widgets which fill gaps in backend functionality to not add entrypoints to *other* projects in library code: rather have the applications do that so they can pick and choose what they use that way.

> I'm not sure if this is drift from an earlier draft - should this be factory? Where does toolkit come from?

Yes, that's a typo, it should be factory or something similar. `toolkit` comes from force of habit from my work on TraitsUI, where `toolkit` is roughly synonymous with factory/backend.

> I think these fall into the category of "interesting crimes that may be possible but that we probably shouldn't encourage" :-)

I actually came up with better ways of doing both of those, which essentially are "make your own backend entrypoint and populate it however you please with widget entrypoints and then use that backend."

**Comment by freakboy3742** [comment]
> So you want that cache _somehow_ - it doesn't have to be a class that's holding it.

Agreed.
 
> The other part of the API above was to keep the API similar to what is there now, but it doesn't have to be via attribute lookup. Either the widget needs to be given the backend name somehow, or it needs to be given an object which wraps that. My feeling is that the latter may be easier to test and mock, for example, but at its core, as long as you have the cache it's all API details.

I guess this also means that there doesn't need to be a direct correlation between the "registered widget" namespace and the Python package namespace. You could register a `toga.hardware-camera` endpoint, implemented in cocoa with `toga_cocoa.some_detail.hardware.camera`. The person building the widget just needs to define *a* namespace. 

I still think I'd prefer having a single entry point, with the "package namespace" being something that is passed in on usage, rather than having a per-module factory.
 
> Given that you're going to need to do the entrypoint queries anyway, I think the overhead is minimal and you won't win much by having some values hard-coded.
... 
> This proposal does have this lazy-loading advantage: because the imports are only done when the widgets are requested, you're not needing to import potentially heavy-weight widgets like web views if you aren't using them. In fact, because entrypoints can be split across packages, backend implementations can be split so you don't even have to include the web view in your app's runtime if you don't need it.

Thinking about it some more, I suspect you're correct that there won't be a performance overhead (or, at least, no large-scale performance degradation).
 
> > The one possible hole I can see is what happens if two modules define an entry point for the same widget? 
...
> So if you had two implementations of a widget (say two implementations of a missing core widget from a backend), then there'll be two values for that widget, which we _can_ detect and warn, and then pick one (we have no way of knowing if one or the other is preferable, although there may be some heuristics we could come up with).
...
> This sort of collision is inherent in any sort of plugin system. The best you can do is have good heuristics and a work-around for when the heuristics fail.

Yeah - I don't think this is something that is worth getting *overly* concerned about; just something to keep in mind as an edge case that requires a warning at a minimum.
 
---
At this point, I can't see any major issues with the design as proposed; I think it is worth at least exploring with some initial prototypes.

**Comment by corranwebster** [comment]
> I guess this also means that there doesn't need to be a direct correlation between the "registered widget" namespace and the Python package namespace. You could register a `toga.hardware-camera` endpoint, implemented in cocoa with `toga_cocoa.some_detail.hardware.camera`. The person building the widget just needs to define _a_ namespace.
> 
> I still think I'd prefer having a single entry point, with the "package namespace" being something that is passed in on usage, rather than having a per-module factory.

I'm coming back to this to look at implementing this now that the Qt backend is reasonably complete, and I'm not quite sure what you mean by this last sentence.

Is it just that instead of:
``` python
from toga import Widget
from my_widgets.factory import my_factory

class BitmapView(Widget):

    def create(self):
        self._impl = my_factory.BitmapView(self)
```
(where `my_factory` is something like `Factory("my_widgets")`), you'd prefer something like:
``` python
from toga import Widget
from toga.platform import get_factory

class BitmapView(Widget):

    def create(self):
        self._impl = get_factory("my_widgets").BitmapView(self)
```
with maybe `get_factory()` defaulting to getting the factory for the core widgets by default.

**Comment by freakboy3742** [comment]
> I'm coming back to this to look at implementing this now that the Qt backend is reasonably complete, and I'm not quite sure what you mean by this last sentence.
...
> with maybe `get_factory()` defaulting to getting the factory for the core widgets by default.

That was the idea, yes. Rather than every plugin having it's own registry, and/or core having a "factory factory" for producing an independent registry for each plugin, there would only be one cache/registry, provided by Toga's core, providing a single source of registration truth. All widgets would look to core to retrieve implementations, rather than to the "root" module.
